File: kotlin/bc-envelope/src/test/kotlin/com/blockchaincommons/bcenvelope/EncapsulationTest.kt
package com.blockchaincommons.bcenvelope

import com.blockchaincommons.bccomponents.EncapsulationScheme
// registerTags() from bc-envelope package initializes GlobalFormatContext
import kotlin.test.Test
import kotlin.test.assertEquals

class EncapsulationTest {

    private fun testScheme(scheme: EncapsulationScheme) {
        val (privateKey, publicKey) = scheme.keypair()
        val envelope = helloEnvelope()
        val encryptedEnvelope = envelope
            .encryptToRecipient(publicKey)
            .checkEncoding()
        val decryptedEnvelope = encryptedEnvelope
            .decryptToRecipient(privateKey)
        assertEquals(
            envelope.structuralDigest(),
            decryptedEnvelope.structuralDigest()
        )
    }

    @Test
    fun testEncapsulation() {
        registerTags()

        testScheme(EncapsulationScheme.X25519)
        testScheme(EncapsulationScheme.MLKEM512)
        testScheme(EncapsulationScheme.MLKEM768)
        testScheme(EncapsulationScheme.MLKEM1024)
    }
}

File: go/bcenvelope/encapsulation_test.go
package bcenvelope

import (
	"testing"

	bccomponents "github.com/nickel-blockchaincommons/bccomponents-go"
)

func testEncapsulationScheme(t *testing.T, scheme bccomponents.EncapsulationScheme) {
	t.Helper()

	privateKey, publicKey, err := bccomponents.EncapsulationKeypair(scheme)
	if err != nil {
		t.Fatalf("keypair generation failed for %v: %v", scheme, err)
	}

	envelope := helloEnvelope()
	encryptedEnvelope := envelope.EncryptToRecipient(publicKey)
	encryptedEnvelope = checkEncoding(t, encryptedEnvelope)

	decryptedEnvelope, err := encryptedEnvelope.DecryptToRecipient(privateKey)
	if err != nil {
		t.Fatalf("decrypt failed for %v: %v", scheme, err)
	}
	if !envelope.StructuralDigest().Equal(decryptedEnvelope.StructuralDigest()) {
		t.Fatalf("structural digest mismatch for %v", scheme)
	}
}

func TestEncapsulation(t *testing.T) {
	bccomponents.RegisterTags()

	testEncapsulationScheme(t, bccomponents.EncapsulationX25519)
	testEncapsulationScheme(t, bccomponents.EncapsulationMLKEM512)
	testEncapsulationScheme(t, bccomponents.EncapsulationMLKEM768)
	testEncapsulationScheme(t, bccomponents.EncapsulationMLKEM1024)
}
